Second post on becoming a Jenkins contributor (newbie-friendly ticket solved) (#2300)
jenkins-infra/jenkins.io: content/blog/2019/05/2019-05-30-becoming-contributor-newbie-tickets.adoc

---
layout: post
title: "The journey of becoming a Jenkins contributor: Newbie-friendly tickets"
tags:
- jenkins
- community
- developer
- contributing
- newcomer

> author: romenrg

Reviewer: You can improve social media references to this blog post by adding a few more entries here.  Refer to the new content in "[Adding a blog post](https://github.com/jenkins-infra/jenkins.io/blob/master/CONTRIBUTING.adoc#adding-a-blog-post)" for more details.

```suggestion
author: romenrg
description: "Fixing a newbie friendly Jenkins issue"
opengraph:
  image: /images/post-images/2019-05-30-becoming-contributor/pr-merged.png
```

---

Two months ago I published an https://jenkins.io/blog/2019/03/29/becoming-contributor-intro/[introductory article] on
the journey of becoming a Jenkins contributor. In that first article, the jenkins.io site was reviewed, learning about
the multiple ways in which we can participate and contribute. Then, a first--basic--contribution I made to the site
repository was described.

Now, in this new article we will be exploring more advanced contributions, commiting code to the actual Jenkins core.

== Getting started with tickets and processes

=== Beginners guide to contributing and Jenkins Jira

> The https://wiki.jenkins.io/display/JENKINS/Beginners+Guide+to+Contributing[beginners guide to contributing to Jenkins]

Author: I've added a reference to the developer pages in the same paragraph, now linking both the new developer pages and the beginners guide to contributing

=== Processes

At this stage, when we have decided to take ownership of a ticket, it’s a good practice to let the rest of the community
know that we are planning to start working on it. We can do so easily, by assigning the ticket to ourselves (see the
“_Assign_” button below the ticket summary).

Assigning the ticket to ourselves in the Jenkins Jira will allow any other contributors to know that we are planning to
take care of the ticket; and in case they are also interested in contributing to it, they will know who to reach if they
want to coordinate work or ask for status. That said, it is worth mentioning that assigning a ticket to yourself does
not mean that other contributors cannot work on it from then onwards. Jenkins is an open-source project and anyone is
welcome to create their own PRs, so anybody can propose their own solution to the ticket. But as you can guess, if the
ticket is assigned to somebody, most people will probably reach the assignee before starting to work on it.

Related to the former, it is also important to bear in mind that we should not postpone work on the ticket for too long

> once we have assigned the ticket to ourselves. Other potential contributors might be discarding the ticket because they

Reviewer: I think you mean "ignoring" or possibly "disregarding" rather than "discarding".

```suggestion
once we have assigned the ticket to ourselves. Other potential contributors might be ignoring the ticket because they
```

see yourself assigned to it.

Once we are about to actually start working on the ticket, it is also a good practice to click the “_Start Progress_”
button. This action will change the status to “_In progress_”, signaling to the community that we are currently working
on this particular ticket.

== Setting up the necessary stuff on our computer

=== Configuring, installing and testing

As described in the https://jenkins.io/blog/2019/03/29/becoming-contributor-intro/[first article of this journey], the
initial step to start contributing to a particular repository is to fork it to our GitHub account, and then clone it to
our local computer.

As usual, in the Jenkins core repository the https://github.com/jenkinsci/jenkins/blob/master/CONTRIBUTING.md[CONTRIBUTING file]
describes the necessary steps to get the repository working locally. This includes installing the necessary development

> tools: Java Development Kit (_http://openjdk.java.net/[OpenJDK] is the recommended choice_), Maven and any IDE supporting

Reviewer: I find it easier to locate a JDK to download from the Adopt OpenJDK download site.  Others in the Jenkins project also seem to prefer that site rather than using openjdk.net.

```suggestion
tools: Java Development Kit (_https://adoptopenjdk.net/[OpenJDK] is the recommended choice_), Maven and any IDE supporting
```

To reproduce the issue locally in our computer, we can either use the local `war` file that we can generate by
https://github.com/jenkinsci/jenkins/blob/master/CONTRIBUTING.md#building-and-debugging[building Jenkins from the source code]
or we can download the https://jenkins.io/download/[latest Jenkins version] available and run it locally. When I worked
on this ticket, the latest available version was 2.172` and, when I built if from the sources I saw version `2.173-SNAPSHOT`,
which was the next version, in which the community was already working on.

In general it is a good idea to reproduce bugs locally, not only to get a better understanding, but also to make sure
they are actual issues. It could always be an issue happening only on the reporter’s end (e.g. some user misconfiguration).
Or it could be a ticket referencing an old issue that has been fixed already. This latest possibility didn’t sound that
strange to me, since the ticket was one month old. It could have been handled by someone else in the meantime, without
noticing the ticket existed. Or the contributor might have forgotten to update the ticket in the issue tracker after the
fix was committed.

> So, for all the reasons above, I went ahead in ran the latest Jenkins version locally. From a terminal, went to the

Reviewer: ```suggestion
So, for all the reasons above, I went ahead in ran the latest Jenkins version locally. From a terminal, I went to the
```

> folder in which the war file was placed, and ran `java -jar jenkins.war`, which starts Jenkins locally on http://localhost:8080.

Reviewer: If you don't want the user to have a clickable link to http://localhost:8080/, then you could consider:

```suggestion
folder in which the war file was placed, and ran `java -jar jenkins.war`, which starts Jenkins locally on \http://localhost:8080.
```

Optional, since you may have intentionally made it a hyperlink for them to click.

Author: Thanks for the suggestion, Mark. In this particular case I was indeed having a link intentionally, to foster curiosity in the reader's end about running Jenkins locally :).

From the home page I navigated to the Plugin Manager (clicking the “_Manage Jenkins_” link in the left hand side and
then selecting “_Manage Plugins_” in the list).

In the Manage Plugins page, the list of plugin updates appears. In my case, since I re-used an old `JENKINS_HOME` I had
from an older installation, several plugins showed up in the list, requiring updates. That allowed me to test the behavior
that was supposed to be failing.

When I clicked on the “Select all” option at the bottom, this is what I got:

.Screenshot showing the error, reproduced locally, after clicking “Select All”
image::/images/post-images/2019-05-30-becoming-contributor/jenkins-plugin-manager-updates-selected-bottom.png[]

As it had been reported in the ticket, the behavior was inconsistent. In a previous version, the behavior of the “_All_”
selector had been changed (with the best intent), aiming to only check the compatible plugins. However, as can be seen
in the screenshot, the behavior was not the expected one. Now, neither “all” nor “only compatible” plugins were being

> selected, since some plugins with compatibility issues were also being checked, unintendedly.

Reviewer: ```suggestion
selected, since some plugins with compatibility issues were also being checked, unintentionally.
```

---
layout: post

> title: "The journey of becoming a Jenkins contributor: Newbie-friendly tickets"

Reviewer: Shorter titles often present better for readers.  Since "becoming" hints that the action is happening now, I think the reader will infer that this is describing a journey.
```suggestion
title: "Becoming a Jenkins contributor: Newbie-friendly tickets"
```

tags:
- jenkins
- community
- developer
- contributing
- newcomer
author: romenrg

> description: "Reviewing the process of fixing an issue in Jenkins core; from ticket to PR to release"

Author: Great suggestion!

opengraph:
  image: /images/post-images/2019-05-30-becoming-contributor/pr-merged.png
---

Two months ago I published an https://jenkins.io/blog/2019/03/29/becoming-contributor-intro/[introductory article] on
the journey of becoming a Jenkins contributor. In that first article, the jenkins.io site was reviewed, learning about
the multiple ways in which we can participate and contribute. Then, a first--basic--contribution I made to the site
repository was described.

> Now, in this new article we will be exploring more advanced contributions, commiting code to the actual Jenkins core.

Reviewer: Spelling fix
```suggestion
Now, in this new article we will be exploring more advanced contributions, committing code to the actual Jenkins core.
```

To reproduce the issue locally in our computer, we can either use the local `war` file that we can generate by
https://github.com/jenkinsci/jenkins/blob/master/CONTRIBUTING.md#building-and-debugging[building Jenkins from the source code]
or we can download the https://jenkins.io/download/[latest Jenkins version] available and run it locally. When I worked
on this ticket, the latest available version was 2.172` and, when I built if from the sources I saw version `2.173-SNAPSHOT`,
which was the next version, in which the community was already working on.

In general it is a good idea to reproduce bugs locally, not only to get a better understanding, but also to make sure
they are actual issues. It could always be an issue happening only on the reporter’s end (e.g. some user misconfiguration).
Or it could be a ticket referencing an old issue that has been fixed already. This latest possibility didn’t sound that
strange to me, since the ticket was one month old. It could have been handled by someone else in the meantime, without
noticing the ticket existed. Or the contributor might have forgotten to update the ticket in the issue tracker after the
fix was committed.

> So, for all the reasons above, I went ahead in ran the latest Jenkins version locally. From a terminal, I went to the

Reviewer: Simplify phrasing
```suggestion
So, for all the reasons above, I ran the latest Jenkins version locally. From a terminal, I went to the
```

folder in which the war file was placed, and ran `java -jar jenkins.war`, which starts Jenkins locally on http://localhost:8080.

From the home page I navigated to the Plugin Manager (clicking the “_Manage Jenkins_” link in the left hand side and
then selecting “_Manage Plugins_” in the list).

> In the Manage Plugins page, the list of plugin updates appears. In my case, since I re-used an old `JENKINS_HOME` I had

Reviewer: Simplify phrasing
```suggestion
In the Manage Plugins page, the list of plugin updates appears. In my case, since I re-used an old `JENKINS_HOME`
```

Eventually, your changes will also reach the https://jenkins.io/download/lts/[Long-term support (LTS) release], which is
a different release line, aimed for more conservative users. This release line gets synced with the weekly release by
picking, every 12 weeks, a relatively recent weekly release as baseline for the new LTS release. In between, intermediate
LTS releases are created only to include important bug fixes, cherry-picked from the weekly releases. New features are
typically delayed until the next baseline for the LTS release is defined.

Regarding the example described in this post, *it was released in Jenkins `2.175`* (weekly release), soon after being
merged. And will probably be included in the next LTS, which should be released next month (June 2019).

== Done!

And that’s it! We have now covered the whole lifecycle of a new proposed change to Jenkins core. We have reviewed the
process from the very beginning, picking a ticket from the Jenkins issue tracker; all the way to the end, having our
change released in a new Jekins version.

If you have never contributed but are willing to do so, I hope this article motivates you to go back to the list of

> `newbie-friendly` tickets, find one that looks interesting to you, and follow the steps described above, until you see

Reviewer: Make the concluding reference to `newbie-friendly tickets` clickable like the first reference is.  As they heed the call to action, they can immediately click the link and identify how they can help.
```suggestion
https://issues.jenkins-ci.org/browse/WEBSITE-625?jql=labels%20%3D%20newbie-friendly[`newbie-friendly` tickets], find one that looks interesting to you, and follow the steps described above, until you see
```
